dbms/src/Debug/MockExecutor/WindowBinder.h

```diff
@@ -14,31 +14,140 @@
 
 #pragma once
 
+#include <Common/Exception.h>
+#include <Common/typeid_cast.h>
 #include <Debug/MockExecutor/ExecutorBinder.h>
+#include <Flash/Coprocessor/ChunkCodec.h>
+#include <Flash/FlashService.h>
+#include <Parsers/ASTFunction.h>
 #include <Parsers/ASTOrderByElement.h>
+#include <tipb/executor.pb.h>
+#include <tipb/expression.pb.h>
+
+#include <memory>
 
 namespace DB::mock
 {
-// true: unbounded, false: not unbounded
-using MockWindowFrameBound = std::tuple<tipb::WindowBoundType, bool, UInt64>;
+// This struct contains something that could help to build the range frame's tipb::Expr
+struct BuildRangeFrameHelper
+{
+    ASTPtr range_aux_func;
+    ContextPtr context;
+};
+
+class MockWindowFrameBound
+{
+public:
+    // Constructor for non-range frame type
+    MockWindowFrameBound(tipb::WindowBoundType bound_type_, bool is_unbounded_, UInt64 offset_)
+        : bound_type(bound_type_)
+        , is_unbounded(is_unbounded_)
+        , offset(offset_)
+    {}
+
+    // Constructor for range frame type
+    MockWindowFrameBound(
+        tipb::WindowBoundType bound_type_,
+        tipb::RangeCmpDataType cmp_data_type_,
+        const BuildRangeFrameHelper & range_frame_helper_)
+        : bound_type(bound_type_)
+        , is_unbounded(false)
+        , offset(0)
+        , range_frame_helper(range_frame_helper_)
+        , cmp_data_type(cmp_data_type_)
+        , range_frame(nullptr)
+    {}
+
+    MockWindowFrameBound(MockWindowFrameBound && bound)
+    {
+        bound_type = bound.bound_type;
+        is_unbounded = bound.is_unbounded;
+        offset = bound.offset;
+        range_frame_helper = bound.range_frame_helper;
+        cmp_data_type = bound.cmp_data_type;
+        range_frame = bound.range_frame;
+
+        bound.offset = 0;
+        bound.range_frame_helper.range_aux_func.reset();
+        bound.range_frame_helper.context.reset();
+        bound.range_frame = nullptr;
+    }
+
+    ~MockWindowFrameBound()
+    {
+        RUNTIME_ASSERT(range_frame == nullptr, "range_frame should not hold data when we are destructed");
+    }
+
+    MockWindowFrameBound(const MockWindowFrameBound &) = default;
+
+    MockWindowFrameBound & operator=(const MockWindowFrameBound & bound) = default;
+
+    tipb::WindowBoundType getBoundType() const { return bound_type; }
+    bool isUnbounded() const { return is_unbounded; }
```

Comment:
Is is_unbounded useful?
Because `std::optional<MockWindowFrameBound>` has been used in `MockWindowFrame`

Reply:
> Is is_unbounded useful? Because `std::optional<MockWindowFrameBound>` has been used in `MockWindowFrame`

Yes, this is useful